File: src/main/java/com/d3bot/events/models/Event.java
package com.d3bot.events.models;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE_TIME;

public record Event(String artist, String location, LocalDateTime dateTime, String url) {

    private static final DateTimeFormatter FORMATTER = ISO_LOCAL_DATE_TIME;

    public String checksum() {
        String input = String.join("|",
                normalise(location),
                normalise(FORMATTER.format(dateTime))
        );
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] hash = md.digest(input.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate checksum", e);
        }
    }

    public String key() {
        return "banquet:event:" + checksum();
    }

    private String normalise(String value) {
        return value.trim().toLowerCase();
    }
}

File: internal/models/event.go
package models

import (
	"crypto/md5"
	"fmt"
	"strings"
	"time"
)

type Event struct {
	Artist   string
	Location string
	DateTime time.Time
	URL      string
}

func (e Event) Checksum() string {
	input := strings.Join([]string{
		normalise(e.Location),
		normalise(e.DateTime.Format("2006-01-02T15:04:05")),
	}, "|")
	return fmt.Sprintf("%x", md5.Sum([]byte(input)))
}

func (e Event) Key() string {
	return "banquet:event:" + e.Checksum()
}

func normalise(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}
